internal/controller/apiextensions/composite/composed.go

```diff
@@ -54,6 +60,224 @@ const (
 	errFmtConnDetailPath = "connection detail of type %q fromFieldPath is not set"
 )
 
+// Annotation keys.
+const (
+	AnnotationKeyCompositionTemplateName = "crossplane.io/composition-template-name"
+)
+
+// SetCompositionTemplateName sets the name of the composition template used to
+// reconcile a composed resource as an annotation.
+func SetCompositionTemplateName(o metav1.Object, name string) {
+	meta.AddAnnotations(o, map[string]string{AnnotationKeyCompositionTemplateName: name})
+}
+
+// GetCompositionTemplateName gets the name of the composition template used to
+// reconcile a composed resource from its annotations.
+func GetCompositionTemplateName(o metav1.Object) string {
+	return o.GetAnnotations()[AnnotationKeyCompositionTemplateName]
+}
```

Comment:
I guess this is subjective, but I kind of prefer it as is. The `resource.Composite` interface has 'critical' getter and setter methods - all the things in spec and status that we couldn't otherwise get at without a type assertion. The interface lets us `GetAnnotations` regardless of what kind of composite we're working with, and this builds on that. This implementation matches the `GetExternalName` pattern, which is also a utility function rather than a method on all managed resources.